### PR TITLE
Add BadDateProcessor

### PR DESCRIPTION
**Pull request by addshore**
This adds BadDateProcessor to list 'bad dates'
This also refactors the main class to be OO and further allows the optional running of processors

### CODE REVIEW

File: java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+public class BadDateProcessor implements EntityDocumentProcessor {
+
+    private BufferedWriter writer1;
+    private BufferedWriter writer2;
+
+    private String gregorianCalendar = "http://www.wikidata.org/entity/Q1985727";
+    private String julianCalendar = "http://www.wikidata.org/entity/Q1985786";
+
+    public BadDateProcessor( BufferedWriter writer1, BufferedWriter writer2 ) throws IOException {
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java]: Inconsistent spacing!

File: java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        this.writer1 = writer1;
+        this.writer1.write("Dates marked as Julian that are more precise than year\n----\n");
+        this.writer2 = writer2;
+        this.writer2.write("Dates marked as gregorian, before 1584\n----\n");
+    }
+
+    @Override
+    public void processItemDocument(ItemDocument item) {
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java]: Is that `@Override` needed?

File: java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                        }
+                    }
+
+                    //List2 - marked as gregorian, before 1584
+                    if(timeValue.getPreferredCalendarModel().equals( this.gregorianCalendar )
+                            && timeValue.getYear() < 1584) {
+                        try {
+                            this.writer2.write(statement.getStatementId() + "\n");
+                        } catch (IOException e) {
+                            System.out.println("Failed to write line to writer2");
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java]: This is kinda odd

File: java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    private BufferedWriter writer1;
+    private BufferedWriter writer2;
+
+    private String gregorianCalendar = "http://www.wikidata.org/entity/Q1985727";
+    private String julianCalendar = "http://www.wikidata.org/entity/Q1985786";
+
+    public BadDateProcessor( BufferedWriter writer1, BufferedWriter writer2 ) throws IOException {
+        this.writer1 = writer1;
+        this.writer1.write("Dates marked as Julian that are more precise than year\n----\n");
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java]: Did you forget to fill this in?

File: java/analyzer/src/main/java/org/wikidata/analyzer/WikidataAnalyzer.java [hunk 5]
```diff
@@ -63,21 +113,28 @@ public static void main(String[] args) throws IOException {
...
+            File mapLabelFile = new File( dataDir.getAbsolutePath() + File.separator + "wdlabel.json");
+            BufferedWriter mapLabelWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(mapLabelFile)));
+            mapGeoData.writeJSONString(mapLabelWriter);
+            mapLabelWriter.close();
+            System.out.println("Writing map graph.json");
+            File mapGraphFile = new File( dataDir.getAbsolutePath() + File.separator + "graph.json");
+            BufferedWriter mapGraphWriter = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(mapGraphFile)));
+            mapGraphData.writeJSONString(mapGraphWriter);
+            mapGraphWriter.close();
+        }
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/WikidataAnalyzer.java]: This is a rather long method, over 50 lines. It is also full of low level details such as strings to output and contains things such as construction logic. Splitting this up a bit would be nice

File: java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void processItemDocument(ItemDocument item) {
+        for (Iterator<Statement> statements = item.getAllStatements(); statements.hasNext(); ) {
+            Statement statement = statements.next();
+            Snak snak = statement.getClaim().getMainSnak();
+            if (snak instanceof ValueSnak) {
+                Value value = ((ValueSnak) snak).getValue();
+                if (value instanceof TimeValue) {
+                    TimeValue timeValue = (TimeValue) value;
```
Comment (JeroenDeDauw) [inline java/analyzer/src/main/java/org/wikidata/analyzer/Processor/BadDateProcessor.java]: I'd extract the body of this if statement out into its of method

**Comment by JeroenDeDauw** [comment]
Can't say much about the design without being more up to speed with wdtk. It's cool you are using wdtk though. OTHO: no tests. aaaaaaaaaaaaaaaaah!

**Comment by addshore** [comment]
*sneeze merge*

**Comment by addshore** [comment]
If I ever touch this thing again I will look at tests!

**Pull request merged**